ql/src/java/org/apache/hadoop/hive/ql/Driver.java

```diff
@@ -139,205 +119,215 @@ public Driver(QueryState queryState, QueryInfo queryInfo, HiveTxnManager txnMana
-    driverTxnHandler.createTxnManager();
-    DriverState.setDriverState(driverState);
-    prepareContext();
-    setQueryId();
+  @Override
+  public CommandProcessorResponse run(String command) throws CommandProcessorException {
+    return run(command, false);
+  }
 
-    if (resetTaskIds) {
-      TaskFactory.resetId();
+  private CommandProcessorResponse run(String command, boolean alreadyCompiled) throws CommandProcessorException {
+    try {
+      runInternal(command, alreadyCompiled);
+      return new CommandProcessorResponse(getSchema(), null);
+    } catch (CommandProcessorException cpe) {
+      processRunException(cpe);
+      throw cpe;
     }
   }
 
-  private void prepareContext() throws CommandProcessorException {
-    if (context != null && context.getExplainAnalyze() != AnalyzeState.RUNNING) {
-      // close the existing ctx etc before compiling a new query, but does not destroy driver
-      closeInProcess(false);
-    }
+  private void runInternal(String command, boolean alreadyCompiled) throws CommandProcessorException {
+    DriverState.setDriverState(driverState);
+    setInitialStateForRun(alreadyCompiled);
 
+    // a flag that helps to set the correct driver state in finally block by tracking if
+    // the method has been returned by an error or not.
+    boolean isFinishedWithError = true;
     try {
-      if (context == null) {
-        context = new Context(driverContext.getConf());
+      HiveDriverRunHookContext hookContext = new HiveDriverRunHookContextImpl(driverContext.getConf(),
+          alreadyCompiled ? context.getCmd() : command);
+      runPreDriverHooks(hookContext);
+
+      if (!alreadyCompiled) {
+        compileInternal(command, true);
+      } else {
+        driverContext.getPlan().setQueryStartTime(driverContext.getQueryDisplay().getQueryStartTime());
       }
-    } catch (IOException e) {
-      throw new CommandProcessorException(e);
-    }
 
-    context.setHiveTxnManager(driverContext.getTxnManager());
-    context.setStatsSource(driverContext.getStatsSource());
-    context.setHDFSCleanup(true);
+      // Reset the PerfLogger so that it doesn't retain any previous values.
+      // Any value from compilation phase can be obtained through the map set in queryDisplay during compilation.
+      PerfLogger perfLogger = SessionState.getPerfLogger(true);
 
-    driverTxnHandler.setContext(context);
-  }
+      // the reason that we set the txn manager for the cxt here is because each query has its own ctx object.
+      // The txn mgr is shared across the same instance of Driver, which can run multiple queries.
+      context.setHiveTxnManager(driverContext.getTxnManager());
 
-  private void setQueryId() {
-    String queryId = Strings.isNullOrEmpty(driverContext.getQueryState().getQueryId()) ?
-        QueryPlan.makeQueryId() : driverContext.getQueryState().getQueryId();
+      DriverUtils.checkInterrupted(driverState, driverContext, "at acquiring the lock.", null, null);
```

Comment:
It might be better message that it was interrupted "before acquiring the locks"
Also getting the locks might take serious amount of time, so it might worth check for interrupt after it too. What do you think